### ISSUE TITLE
When is a MediaStream playing?

### ISSUE THREAD

**Issue by jan-ivar** [w3c/mediacapture-main#615]
Consider setting an empty stream on an element and adding a track later. When is it playing?
```js
video.srcObject = new MediaStream();
video.play();                    // video.currentTime = 0
await wait(5000);                // video.currentTime = ?
video.srcObject.addTrack(track); // video.currentTime = ?
```

[mediacapture-main says](https://w3c.github.io/mediacapture-main/#mediastreams-in-media-elements): *"A MediaStream is not preloadable or seekable and represents a simple, potentially infinite, linear media timeline. The timeline starts at 0 and increments linearly in real time as long as the* ***MediaStream is playing.*** *The timeline does not increment when the playout of the MediaStream is paused."*

What definition of *"playing"* is used? We probably mean [potentially playing](https://html.spec.whatwg.org/multipage/media.html#potentially-playing), a definition the element uses e.g. [here](https://html.spec.whatwg.org/multipage/media.html#media-playback): *"When a media element is potentially playing ... its current playback position must increase monotonically"*.

I wrote a [fiddle](https://jsfiddle.net/jib1/pb0uwa6s/) to check what browsers do in this case, and it's a bit all over the map:

* Firefox 68 fires 1st `timeupdate` w/time `~0.01` on empty stream set.
* Firefox 70 fires 1st `timeupdate` w/time `~0.01` when track is added to stream later.
* Chrome 76 fires 1st `timeupdate` w/time `~177281` when track is added to stream later.¹
* Chrome 78 same
* Safari 12 plays black, Fires 1st and only `timeupdate` w/time `0` when track is added.
* Safari 13 plays black. No events.
* Edge 44 `Error: Unexpected call to method or property access` on empty stream set.

Using the [potentially playing](https://html.spec.whatwg.org/multipage/media.html#potentially-playing) definition would mean Firefox 70 and Chrome¹ have the correct behavior, I think. That definition relies on `readyState` being `HAVE_ENOUGH_DATA` which I observe being the case when `timeupdate` fires.

---
<sub>1. Chrome's `currentTime` jump from 0 to thousands of seconds seems like a bug orthogonal to this spec question.</sub>

**Comment by Pehrsons** [comment]
I updated the [fiddle](https://jsfiddle.net/pehrsons/5vg8zaf4/) to first add a video track that hadn't seen any frames yet (to have a track but not advance the media element's readyState). It shows Firefox 70 not adhering to "potentially playing" as it fires the first `timeupdate` after the frameless track is added. Chrome fires its first `timeupdate` after `canplaythrough` a la "potentially playing".

**Comment by jan-ivar** [comment]
I think we should fix Firefox there.

**Comment by Pehrsons** [comment]
[Bug 1577505](https://bugzilla.mozilla.org/show_bug.cgi?id=1577505) filed.

**Comment by guest271314** [comment]
When `await play()` is used at Chromium 78  `timeupdate` event is not dispatched. `play` event then `waiting` event are dispatched.

Firefox 68 and Nightly 70 throw `MediaStreamError` 
​
```
constraint: ""
​
message: "Starting video failed"
​
name: "AbortError"
​
stack: ""
```

**Comment by guest271314** [comment]
@Pehrsons When `await play()` is used `timeupdate` is not reached https://jsfiddle.net/4qu2dy1o/.

**Comment by guest271314** [comment]
Interestingly this line 

`console.log((await video.play()) instanceof Promise, "Is play() a Promise?");
`
does not have any output.

**Comment by Pehrsons** [comment]
@guest271314 If you find implementation bugs, please file them directly with the implementations. Chromium: crbug.com, Firefox: bugzilla.mozilla.org

**Comment by guest271314** [comment]
@Pehrsons https://bugzilla.mozilla.org/show_bug.cgi?id=1582380, https://bugs.chromium.org/p/chromium/issues/detail?id=1005654.

**Comment by guest271314** [comment]
One interesting observable difference between Chromium, Chrome and Firefox, Nightly re when a `MediaStream` is playing is when a `MediaStream` is set as `srcObject` of an `HTMLMediaElement` and the pause control is clicked 

**Chromium**

`currentTime` continues to increment during the paused state

**Firefox**

`currentTime` does not increment during the paused state

**Comment by Pehrsons** [comment]
I'll add a testcase for that over in https://bugzilla.mozilla.org/show_bug.cgi?id=1588507.

**Comment by guest271314** [comment]
@Pehrsons It is not clear - from perspective here - if the relevant specification (this or capture from element) states what should occur as to `currentTime`  when the pause control is activated by the user. What is your reading of the specification(s) for that case?

**Comment by Pehrsons** [comment]
It is pretty clear to me (see below). mediacapture-fromelement has no say here.

[mediacapture-main](https://w3c.github.io/mediacapture-main/#mediastreams-in-media-elements) says this about currentTime when srcObject is a MediaStream (in the table, for the currentTime row):

> The initial value is 0 and the values increments linearly in real time whenever the stream is playing.

I cannot imagine the intention of "playing" to include when the media element is paused, even if "playing" isn't well defined in the media element spec. Potentially playing, if that's what we end up changing mediacapture-main to, explicitly says that it's only true [when `paused` is false](https://html.spec.whatwg.org/multipage/media.html#potentially-playing).

**Comment by guest271314** [comment]
This is the code used https://github.com/w3c/mediacapture-main/issues/628#issuecomment-541337186.

Technically, the media stream is still playing even when the specific media element where `srcObject` is set to the `MediaStream` is paused. Consider more than one media element having the same `MediaStream` as source. Simply because one media element is paused does not translate into the `MediaStream` itself not playing.

**Comment by guest271314** [comment]
@Pehrsons The linked code was intended to be an infinite media stream using the available `timeupdate` event to to push a frame to be created and added to the stream. 

Another difference between Mozilla Firefox, Nightly and Chromium, Chrome implementations  relevant to playback

Chromium requires execution of code which creates and requests a frame at the first `unmute` event of video `MediaStreamTrack` for playback does not proceed, `currentTime` is `0`, `play`, `playing` events of `HTMLMediaElement` are dispatched, then `mute` event of `MediaStreamTrack` is dispatched. 

At Mozilla the call to `handleTimeupdate()` can be commented, playback continues, `currentTime` increments, `timeupdate` event is dispatched.

```
    canvasTrack.onmute = e => {
      console.log(e.type);
      handleTimeupdate({
        target: Object.assign(
          video, e.target.getSettings()
        )
      });
    };
```

**Comment by Pehrsons** [comment]
> This is the code used [#628 (comment)](https://github.com/w3c/mediacapture-main/issues/628#issuecomment-541337186).
> 
> Technically, the media stream is still playing even when the specific media element where `srcObject` is set to the `MediaStream` is paused. Consider more than one media element having the same `MediaStream` as source. Simply because one media element is paused does not translate into the `MediaStream` itself not playing.

There is no concept of playing for a MediaStream. There is no concept of playing for a MediaStreamTrack. There is a concept of playing for HTMLMediaElement.

Yes, data is flowing under MediaStreamTracks even when no HTMLMediaElement is playing them. That's because there is no buffering of data (in MediaStreamTrack *or* HTMLMediaElement), and regardless of them playing or not playing in any HTMLMediaElement.

TL;DR Chrome's implementation has a bug they need to fix to become spec compliant.

> (Which would mean Mozilla behaviour is not specification compliant - though interestingn and potentially useful. 

The specs trump your reasoning, and with [bug 1577505](https://bugzilla.mozilla.org/show_bug.cgi?id=1577505) fixed, and this issue resolved so that "playing" is changed to "potentially playing", Firefox is spec compliant wrt currentTime in a HTMLMediaElement playing a MediaStream.

> E.g., also technically a canvas media stream could be considered "paused" when no images are being drawn onto the canvas - which brings into purview the different ways `mute` and `unmute` are dispatched by Chromium and Mozilla. At Chromium when paused is clicked `mute` and `unmute` events can toggle every few seconds. Unfortunately lost some code during recent experiments when the OS became unresponsive - though the linked code used should be modifiable to the degree necessary to test various cases)

No spec says a CanvasCaptureMediaStreamTrack should fire "mute" when nothing is drawn to its canvas. mediacapture-fromelement says to fire "mute" on CanvasCaptureMediaStreamTrack when its canvas' origin-clean flag becomes false. Firefox doesn't implement that, and it sounds like Chrome implements a bit too much.

Further than that I'm not going to comment on or debug your code.


> Another difference between Mozilla Firefox, Nightly and Chromium, Chrome implementations relevant to playback

This issue does not handle differences in playback of MediaStreams in different implementations. It discusses the meaning of "playing" in this spec. Let's keep it at that.

**Comment by guest271314** [comment]
@Pehrsons 

> There is no concept of playing for a MediaStream.

The title of this issue is 

> When is a MediaStream playing?

> The specs trump your reasoning, and with bug 1577505 fixed, and this issue resolved so that "playing" is changed to "potentially playing", Firefox is spec compliant wrt currentTime in a HTMLMediaElement playing a MediaStream.

Not sure what you are referring to specifically here. At Firefox when the first frame is received `timeupdate` is dispatched which apparently leads to `timeupdate` being dispatched at `HTMLMediaElement`. Chromium does not dispatch `timeupdate` when the first frame is received evidenced by `currentTime` remaining at `0`.

> Further than that I'm not going to comment on or debug your code.

Did not ask for the code to be debugged. Just happened to be experimenting with mechanisms to push-pull, pull-push, or "drive" a `MediaStreamTrack` of kind video using `mute`, `unmute` and if necessary `timeupdate` of `HTMLMediaElement`, thus posted the code to contribute to the testing already being done, FWIW.

**Comment by guest271314** [comment]
@Pehrsons 

> Firefox is spec compliant wrt currentTime in a HTMLMediaElement playing a MediaStream.

Can you cite and explain the specification reasoning behind `currentTime` not incrementing at Firefox when the pause control is activated by the user and throughout the paused state - even though the `MediaStreamTrack` is still "playing" (not `mute`, `enabled`, and is receiving frames)? E.g., when a `<video>` with `MediaStream` set at `srcObject` is paused by user activation, within the `pause` event handler a second `<video>` is created and set to the same `MediaStream` the `currentTime` of the second `<video>` increments, is not paused, however, when the play control is clicked again at the first `<video>` `currentTime` re-commences at the same `currentTime` the pause occurred, providing the apparently impression that the `MediaStream` was paused at `currentTime`.

**Comment by Pehrsons** [comment]
See https://github.com/w3c/mediacapture-main/issues/615#issuecomment-541727689.

Note that the only references to "playing" in mediacapture-main are in relation to HTMLMediaElement, which is why I draw the conclusion that MediaStreams and MediaStreamTracks themselves don't have a concept of playing. The intent of mediacapture-main seems to be to refer to a HTMLMediaElement playing the MediaStream.

**Comment by guidou** [comment]
I agree with @Pehrsons interpretation. We should change mediacapture-main to say "potentially playing" so that it better aligns with the media element spec.
Filed crbug.com/1014541 to track the bug in Chrome.

**Comment by guest271314** [comment]
@Pehrsons In the midst of attempting to test playback after updating Firefox and Nightly all of a sudden `getUserMedia({video:true})` "stopped working" throws an error or exception though `getUserMedia({audio:true})` does not https://bugzilla.mozilla.org/show_bug.cgi?id=1588939. Not sure if the issue is related to https://stackoverflow.com/questions/57419954/navigator-mediadevices-getusermedia-notallowederror ? The context is `https:`. (FWIW re https://github.com/WICG/speech-api/issues/70#issuecomment-542272606 in spite of https://github.com/WICG/speech-api/issues/67#issuecomment-539060815 am banned from WICG)
